### PR TITLE
Events testing methods

### PR DESCRIPTION
**Pull request by kubaplas**

### CODE REVIEW

File: odra-core/src/host_env.rs [hunk 1]
```diff
@@ -131,30 +131,85 @@ impl HostEnv {
...
+
+    pub fn event_names(&self) -> Vec<String> {
+        let mut event_names = vec![];
+        self.events_count
+            .borrow()
+            .iter()
+            .for_each(|(contract_address, _)| {
+                event_names.append(&mut self.contract_event_names(contract_address));
+            });
+        event_names
```
Comment (kpob) [inline odra-core/src/host_env.rs]: Could be written without a tmp vec
```suggestion
        self.events_count
            .borrow()
            .iter()
            .map(|(contract_address, _)| self.contract_event_names(contract_address))
            .flatten() // to flateen Vec<Vec<String>>
            .collect()
```

File: examples2/src/erc20.rs [hunk 2]
```diff
@@ -913,6 +932,95 @@ mod tests {
...
+            .contains(&"Transfer".to_string()));
+        // or
+        assert!(env.last_call().emitted("Transfer"));
+
+        // To check whole event:
+        // TODO: change it to hopefully_emitted.into()
+        assert!(env
+            .last_call()
+            .events()
+            .contains(&Bytes::from(emitted_in_second_call.to_bytes().unwrap())));
```
Comment (kpob) [inline examples2/src/erc20.rs]: I would prefer to pass just an event and do the conversion under the hood.

**Review by kpob — APPROVED** [review]

File: odra-core/src/host_env.rs (hunk 1 above)
Comment (kubaplas) [reply odra-core/src/host_env.rs]: Done

File: examples2/src/erc20.rs (hunk 2 above)
Comment (kubaplas) [reply examples2/src/erc20.rs]: Line below that is a solution

File: examples2/src/erc20.rs [hunk 3]
```diff
@@ -626,6 +623,28 @@ mod __erc20_test_parts {
...
+            self.env.contract_event_names(&self.address)
+        }
+
+        pub fn emitted(&self, event_name: &str) -> bool {
+            self.event_names().contains(&String::from(event_name))
+        }
+
+        pub fn last_call(&self) -> CallResult {
+            let mut last_call = self.env.last_call();
+            last_call.events = last_call
```
Comment (zie1ony) [inline examples2/src/erc20.rs]: Move this logic into `CallResult`.

File: odra-core/src/call_result.rs [hunk 4]
```diff
@@ -34,15 +34,48 @@ impl CallResult {
...
+        self.events.values().for_each(|val| {
+            val.iter().for_each(|bytes| {
+                event_names.push(HostEnv::extract_event_name(bytes).unwrap());
+            })
+        });
+
+        event_names
+    }
+
+    pub fn events(&self) -> Vec<Bytes> {
```
Comment (zie1ony) [inline odra-core/src/call_result.rs]: I'd remove this method and leave only `contract_events`.

File: odra-core/src/call_result.rs [hunk 5]
```diff
@@ -34,15 +34,48 @@ impl CallResult {
...
+    pub fn gas_used(&self) -> u64 {
         self.gas_used
     }
 
-    pub fn get_contract_address(&self) -> Address {
+    pub fn contract_address(&self) -> Address {
         self.contract_address
     }
+
+    pub fn event_names(&self) -> Vec<String> {
```
Comment (zie1ony) [inline odra-core/src/call_result.rs]: I'd make this to `event_names(&self, contract: &Address)`
List of all event names from all contracts, without the origin is not really useful.

File: odra-core/src/call_result.rs [hunk 6]
```diff
@@ -34,15 +34,48 @@ impl CallResult {
...
+        });
+
+        events
+    }
+
+    pub fn contract_events(&self, contract_address: &Address) -> Vec<Bytes> {
+        self.events.get(contract_address).unwrap_or(&vec![]).clone()
+    }
+
+    pub fn emitted(&self, event_name: &str) -> bool {
```
Comment (zie1ony) [inline odra-core/src/call_result.rs:114]: Same as `event_names`. Add `contract: &Address`.

File: odra-core/src/call_result.rs [hunk 7]
```diff
@@ -34,15 +34,48 @@ impl CallResult {
...
+        self.event_names().contains(&event_name.to_string())
+    }
+
+    pub fn emitted_event<T: ToBytes + EventInstance>(&self, event: &T) -> bool {
```
Comment (zie1ony) [inline odra-core/src/call_result.rs:118]: Same as `event_names`. Add `contract: &Address`.

File: odra-core/src/host_env.rs [hunk 8]
```diff
@@ -131,30 +131,85 @@ impl HostEnv {
...
         self.last_call_result.borrow().clone().unwrap().clone()
     }
 }
 
 impl HostEnv {
     /// Returns the name of the passed event
-    fn extract_event_name(bytes: &[u8]) -> Result<String, EventError> {
-        let name = FromBytes::from_bytes(bytes).map_err(|_| EventError::CouldntExtractName)?;
-        Ok(name.0)
+    pub fn extract_event_name(bytes: &[u8]) -> Result<String, EventError> {
```
Comment (zie1ony) [inline odra-core/src/host_env.rs]: It should be in `odra-utils`.

**Review by zie1ony — CHANGES_REQUESTED** [review]

File: odra-core/src/host_env.rs (hunk 8 above)
Comment (kubaplas) [reply odra-core/src/host_env.rs]: It depends on casper_contract and odra_core, so it cannot be there.

File: examples2/src/erc20.rs [hunk 9]
```diff
@@ -454,6 +454,7 @@ impl Erc20ContractRef {
 #[cfg(not(target_arch = "wasm32"))]
 mod __erc20_test_parts {
     use crate::erc20::Erc20;
+    use alloc::boxed::Box;
```
Comment (zie1ony) [inline examples2/src/erc20.rs]: ?

File: odra-core/src/host_env.rs (hunk 8 above)
Comment (zie1ony) [reply odra-core/src/host_env.rs]: make it non public

**Review by zie1ony — APPROVED** [review]

**Pull request merged**